execution/gethexec/sync_monitor.go

```diff
@@ -0,0 +1,106 @@
+package gethexec
+
+import (
+	"context"
+
+	"github.com/offchainlabs/nitro/execution"
+	"github.com/pkg/errors"
+	flag "github.com/spf13/pflag"
+)
+
+type SyncMonitorConfig struct {
+	SafeBlockWaitForBlockValidator      bool `koanf:"safe-block-wait-for-block-validator"`
+	FinalizedBlockWaitForBlockValidator bool `koanf:"finalized-block-wait-for-block-validator"`
+}
+
+var DefaultSyncMonitorConfig = SyncMonitorConfig{
+	SafeBlockWaitForBlockValidator:      false,
+	FinalizedBlockWaitForBlockValidator: false,
+}
+
+func SyncMonitorConfigAddOptions(prefix string, f *flag.FlagSet) {
+	f.Bool(prefix+".safe-block-wait-for-block-validator", DefaultSyncMonitorConfig.SafeBlockWaitForBlockValidator, "wait for block validator to complete before returning safe block number")
+	f.Bool(prefix+".finalized-block-wait-for-block-validator", DefaultSyncMonitorConfig.FinalizedBlockWaitForBlockValidator, "wait for block validator to complete before returning finalized block number")
+}
+
+type SyncMonitor struct {
+	config    *SyncMonitorConfig
+	consensus execution.ConsensusInfo
+	exec      *ExecutionEngine
+}
+
+func NewSyncMonitor(config *SyncMonitorConfig, exec *ExecutionEngine) *SyncMonitor {
+	return &SyncMonitor{
+		config: config,
+		exec:   exec,
+	}
+}
+
+func (s *SyncMonitor) SyncProgressMap() map[string]interface{} {
+	res := s.consensus.SyncProgressMap()
+	consensusSyncTarget := s.consensus.SyncTargetMessageCount()
+
+	built, err := s.exec.HeadMessageNumber()
+	if err != nil {
+		res["headMsgNumberError"] = err
+	}
+
+	if built+1 >= consensusSyncTarget && len(res) == 0 {
```

Comment:
Not that it matters too much, but why the +1 here? I think since we fetch the latest built after we fetch he target we don't need it.

Reply:
Also, shouldn't this if statement include `err == nil`?

Reply:
+1 is because exec usually (including in HeadMessageNumber) uses message number, and consensus usually (including SyncTargetMessageCount) works with message count, which is +1. It's confusing but trying to switch everything to message number makes it hard to express uninitialized state with no messages, and using message count everywhere requires always adding >0 test.. I try to keep with the division for consensus vs. execution and the naming of number/index vs. count

Reply:
here:
 err != nil ==means==> len(res) != 0
 but in the new version it is a required test.